source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.cc

```diff
@@ -201,15 +217,36 @@ X509_STORE* SPIFFEValidator::getTrustBundleStore(X509* leaf_cert) {
 bool SPIFFEValidator::certificatePrecheck(X509* leaf_cert) {
   // Check basic constrains and key usage.
   // https://github.com/spiffe/spiffe/blob/master/standards/X509-SVID.md#52-leaf-validation
-  auto ext = X509_get_extension_flags(leaf_cert);
+  const auto ext = X509_get_extension_flags(leaf_cert);
   if (ext & EXFLAG_CA) {
     return false;
   }
 
-  auto us = X509_get_key_usage(leaf_cert);
+  const auto us = X509_get_key_usage(leaf_cert);
   return (us & (KU_CRL_SIGN | KU_KEY_CERT_SIGN)) == 0;
 }
 
+bool SPIFFEValidator::matchSubjectAltName(X509& leaf_cert) {
+  bssl::UniquePtr<GENERAL_NAMES> san_names(static_cast<GENERAL_NAMES*>(
+      X509_get_ext_d2i(&leaf_cert, NID_subject_alt_name, nullptr, nullptr)));
+  // We must not have san_names == nullptr here because this function is called after the
+  // SPIFFE cert validation algorithm succeeded, which requires exactly one URI SAN in the leaf
+  // cert.
+  assert(san_names != nullptr);
+
+  // Only match against URI SAN since SPIFFE specification does not restrict values in other SAN
+  // types. See the discussion: https://github.com/envoyproxy/envoy/issues/15392
+  for (const GENERAL_NAME* general_name : san_names.get()) {
+    const std::string san = Utility::generalNameAsString(general_name);
+    for (auto& config_san_matcher : subject_alt_name_matchers_) {
+      if (general_name->type == GEN_URI && config_san_matcher.match(san)) {
+        return true;
+      }
+    }
```

Comment:
Very slight nit on the order of operations here. We can filter by GEN_URI immediately instead of after going through the trouble of extracting the general name as a string and iterating over each matcher.
```suggestion
    if (general_name->type == GEN_URI) { 
        const std::string san = Utility::generalNameAsString(general_name);
        for (auto& config_san_matcher : subject_alt_name_matchers_) {
            if (config_san_matcher.match(san)) {
                return true;
          }
     }

Reply:
Yeah, right. Thanks!

Reply:
fixed https://github.com/envoyproxy/envoy/pull/15509/commits/95cdfcb0c934b0c395673c07762f60b351caf8d5